runtime/vm/UpcallVMHelpers.cpp

```diff
@@ -114,7 +144,525 @@ native2InterpJavaUpcallD(J9UpcallMetaData *data, void *argsListPointer)
+					 * after returning from the native function in downcall.
+					 */
+					throwOOM = true;
+					goto done;
+				}
+				 /* Push the object on the special frame so as to avoid updating the address on the stack
+				  * by GC when allocating memory for the next pointer/struct of the argument list.
+				  */
+				PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, memAddrObject);
+				break;
+			}
+			case J9_FFI_UPCALL_SIG_TYPE_STRUCT:
+			{
+				I_64 offset = (I_64)(intptr_t)getArgPointer(nativeSig, argsListPointer, argIndex);
+				j9object_t memSegmtObject = createMemSegmentObject(data, offset, sigArray[argIndex].sizeInByte, isCurThrdAllocated, sessionOrScopeObject);
+				if (NULL == memSegmtObject) {
+					/* The OOM exception set in createMemSegmentObject() will be thrown in the interpreter
+					 * after returning from the native function in downcall.
+					 */
+					throwOOM = true;
+					goto done;
+				}
+				 /* Push the object on the special frame so as to avoid updating the address on the stack
+				  * by GC when allocating memory for the next pointer/struct of the argument list.
+				  */
+				PUSH_OBJECT_IN_SPECIAL_FRAME(currentThread, memSegmtObject);
+				break;
+			}
+			default:
+				Assert_VM_unreachable();
+				break;
+			}
+		}
+		/* Only restore the literals and keep the unchanged arguments on the stack which
+		 * are passed over to native2InterpreterTransition() in the interpreter so as to
+		 * invoke the upcall method handle.
+		 */
+		currentThread->literals = thrLiterals;
+
+		/* Place mhMetaData as the return value to native2InterpreterTransition() when calling into
+		 * the interpreter so as to set the invoke cache array (MemberName and appendix)
+		 * before invoking the target handle in upcall.
+		 */
+		currentThread->returnValue = J9_BCLOOP_N2I_TRANSITION;
+		currentThread->returnValue2 = (UDATA)data;
+		c_cInterpreter(currentThread);
+
+done:
+		restoreCallInFrameHelper(currentThread);
+	}
+	VM_VMAccess::inlineExitVMToJNI(currentThread);
+
+	if (!throwOOM) {
+		returnStorage = currentThread->returnValue;
+		convertUpcallReturnValue(data, returnType, &returnStorage);
+	}
+
+	/* Release the J9VMThread if allocated locally for the native thread */
+	if (isCurThrdAllocated) {
+		threadCleanup(currentThread, false);
+		currentThread = NULL;
+	}
+
+doneAndExit:
+	return returnStorage;
+}
+
+/**
+ * @brief Get a J9VMThread whether it is the current thread or a newly created thread if doesn't exist
+ *
+ * Note:
+ * The function is to handle the situation when there is no Java thread for the current native thread
+ * which is directly created in native code without a Java thread attached to it.
+ *
+ * @param vm a pointer to J9UpcallMetaData
+ * @param isCurThrdAllocated a pointer to a flag indicating whether the current thread is allocated locally
+ * @return a pointer to J9VMThread
+ */
+static J9VMThread *
+getCurrentThread(J9UpcallMetaData *data, bool *isCurThrdAllocated)
```

Comment:
I think there are a bunch of things missing here like: threadname, threadobject, OOM error instance. Also thread attributes for the omr thread

Reply:
also need to check if it triggers jvmtihooks

Reply:
> I think there are a bunch of things missing here like: threadname, threadobject, OOM error instance. Also thread attributes for the omr thread


If the thread is an async mode, it means the thread is created in the native function, which is different from the original thread in downcall. Under such circumstance, we need a way to differentiate this type of thread from other thread in the sync mode (the thread in downcall & upcall is the same one) given such thread is run within a shared session/scope, which means the owner thread for the session/scope is null (e.g. `session.ownerThread()` is null in the case of the shared session in java). That being said, the `threadobject` for this type of thread must be `null` (I suspect whether `threadname` is still required in such case).

Reply:
I will need to check other things including OOM error instance, omr thread attributes and jvmtihooks to see how they are handled in such case.

Reply:
Im not sure that you can use the downcall thread. Its possible that one could perform a dwoncall, then pass the function pointer (upcall) thunk to another native thread (thread 2)via a global then return from the downcall. The other native thread could then invoke the upcall and you would end up with two native threads with the same vmthread.

Reply:
Does RI allow an arbritary native thread to attach to the JVM via an upcall? If so, what happens if one calls `Thread:currentThread` in the upcall? Does that thread have a name? Threadgroup? etc.

Reply:
I've updated the code (and verified locally on pLinux with our tests and jtreg related test suites) following the implementation of `attachCurrentThread()` at https://github.com/eclipse-openj9/openj9/blob/master/runtime/vm/jniinv.c#L1130 , which covers everything required for a new VMThread with the complete settings in code:
1)`attachThreadWithCategory()` handles thread attributes for the newly created omr thread.
2)`internalAttachCurrentThread()` handles everything for the newly created VMThread, including `threadname`, `threadobject`, threadgroup, jvmtihook event, etc.
3) the OOM errors are set to the downcall thread to bring it back to the interpreter in downcall given the locally created current thread will be cleaned up once the dispatcher exists; otherwise, the current thread is the same thread as the downcall thread.

Note:
1) There is no need to call `getVMThreadFromOMRThread()` as the OMR thread is newly created and the underlying native thread is created in the native function outside of VM. 
2) We can't directly call `AttachCurrentThread()` as we need to set the private flag `J9_PRIVATE_FLAGS_FFI_UPCALL_THREAD` to differentiate this type of thread from other threads so as to identify the session/scope to be set in code.

Reply:
> I think these questions pretty much answer why threadName and threadObject are required in these situations to avoid unexpected results (e.g. Thread:currentThread() returns null if threadObject is not set to the new VM thread)

I'm still curious as to what happens with RI if a native (non-Java) thread calls in and dynamically creates a new JVM thread then call `Thread::currentThread`?

Reply:
I slightly modified the jteg test suite (which previously failed due to the missing the native thread in this case and this fix originally intended for the native thread was to resolve) by adding pieces of printing messages for `Thread.currentThread()` in downcall and the upcall methods at https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/openj9/test/jdk/java/foreign/TestUpcallAsync.java to verify on both RI and our build:
```
public class TestUpcallAsync extends TestUpcallBase {

    @Test(dataProvider="functions", dataProviderClass=CallGeneratorHelper.class)
    public void testUpcallsAsync(int count, String fName, Ret ret, List<ParamType> paramTypes, List<StructFieldType> fields) throws Throwable {
...
        try (MemorySession session = MemorySession.openShared()) {
...
------> System.out.println("testUpcallsAsync: Thread.currentThread() = " + Thread.currentThread());
            MethodHandle mh = downcallHandle(ABI, addr, allocator, descriptor);
```
and https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/openj9/test/jdk/java/foreign/TestUpcallBase.java (which includes the upcall  methods)
```
    static Object passAndSave(Object[] o, AtomicReference<Object[]> ref, int retArg) {
---> System.out.println("TestUpcallBase: passAndSave(): Thread.currentThread() = " + Thread.currentThread());
...
        return o[retArg];
    }

    static void dummy() {
---> System.out.println("TestUpcallBase: dummy(): Thread.currentThread() = " + Thread.currentThread()); 
        //do nothing
    }
```
Specifically, the upcall is triggered in native in a thread separately created in native at https://github.com/ibmruntimes/openj9-openjdk-jdk19/blob/openj9/test/jdk/java/foreign/libAsyncInvokers.cpp
```
e.g.
template<typename CB>
void launch_v(CB cb) {
    std::thread thrd(cb); <-----------
    thrd.join();
}

template<typename O, typename CB>
void start(O& out, CB cb) {
    out = cb();
}

template<typename O, typename CB>
O launch(CB cb) {
    O result;
    std::thread thrd(&start<O, CB>, std::ref(result), cb); <---------
    thrd.join();
    return result;
}

extern "C" {
EXPORT void call_async_V(void (*cb)(void)) { launch_v(cb); }
EXPORT int call_async_I(int (*cb)(void)) { return launch<int>(cb); }
...
```

[1] RI:
```
#section:script_messages
----------messages:(4/232)----------
JDK under test: /home/jincheng/.../jdk19_hotspot_openjdk_x64
openjdk version "19-ea" 2022-09-20
OpenJDK Runtime Environment (build 19-ea+23-1706)
OpenJDK 64-Bit Server VM (build 19-ea+23-1706, mixed mode, sharing)
...
#section:testng
----------messages:(4/406)----------
command: testng -XX:+IgnoreUnrecognizedVMOptions -XX:-VerifyDependencies --enable-native-access=ALL-UNNAMED -Dgenerator.sample.factor=17 TestUpcallAsync
...
----------System.out:(798/76319)----------
config TestUpcallBase.setup(): success
testUpcallsAsync: Thread.currentThread() = Thread[#21,MainThread,5,MainThreadGroup]
TestUpcallBase: dummy(): Thread.currentThread() = Thread[#25,Thread-1,5,main]  <---- created by the native thread in upcall
test TestUpcallAsync.testUpcallsAsync(0, "f0_V__", VOID, [], []): success

testUpcallsAsync: Thread.currentThread() = Thread[#21,MainThread,5,MainThreadGroup]
TestUpcallBase: passAndSave(): Thread.currentThread() = Thread[#26,Thread-2,5,main]  <---- created by the native thread in upcall
test TestUpcallAsync.testUpcallsAsync(17, "f0_V_S_DI", VOID, [STRUCT], [DOUBLE, INT]): success

testUpcallsAsync: Thread.currentThread() = Thread[#21,MainThread,5,MainThreadGroup]
TestUpcallBase: passAndSave(): Thread.currentThread() = Thread[#27,Thread-3,5,main]  <---- created by the native thread in upcall
test TestUpcallAsync.testUpcallsAsync(34, "f0_V_S_IDF", VOID, [STRUCT], [INT, DOUBLE, FLOAT]): success
```
[2] Our build (OpenJ9):
```
#section:script_messages
----------messages:(7/453)----------
JDK under test: /root/.../jdk19_openj9_ppc64le
openjdk version "19-internal" 2022-09-20
OpenJDK Runtime Environment (build 19-internal-adhoc..openj9-openjdk-jdk19-v981)
Eclipse OpenJ9 VM (build jep389_419_ffi_upcall_v4_struct_v2_outofline_mh_v9_fin_v8_5-37c9771, JRE 19 Linux ppc64le-64-Bit Compressed References 20220901_000000 (JIT enabled, AOT enabled)
OpenJ9   - 37c9771
OMR      - 063dc5d
JCL      - 48908d5 based on jdk-19+36)


----------messages:(4/406)----------
command: testng -XX:+IgnoreUnrecognizedVMOptions -XX:-VerifyDependencies --enable-native-access=ALL-UNNAMED -Dgenerator.sample.factor=17 TestUpcallAsync
...
----------System.out:(1023/77904)----------
config TestUpcallBase.setup(): success

testUpcallsAsync: Thread.currentThread() = Thread[#28,MainThread,5,MainThreadGroup]
TestUpcallBase: dummy(): Thread.currentThread() = Thread[#29,Thread-4,5,main] <---- created by the native thread in upcall
test TestUpcallAsync.testUpcallsAsync(0, "f0_V__", VOID, [], []): success

testUpcallsAsync: Thread.currentThread() = Thread[#28,MainThread,5,MainThreadGroup]
TestUpcallBase: passAndSave(): Thread.currentThread() = Thread[#30,Thread-5,5,main] <---- created by the native thread in upcall
test TestUpcallAsync.testUpcallsAsync(17, "f0_V_S_DI", VOID, [STRUCT], [DOUBLE, INT]): success

testUpcallsAsync: Thread.currentThread() = Thread[#28,MainThread,5,MainThreadGroup]
TestUpcallBase: passAndSave(): Thread.currentThread() = Thread[#31,Thread-6,5,main] <---- created by the native thread in upcall
test TestUpcallAsync.testUpcallsAsync(34, "f0_V_S_IDF", VOID, [STRUCT], [INT, DOUBLE, FLOAT]): success

```
The results above indicates the downcall thread is totally different from the upcall method in the async mode (in which the native thread is created separately in native), in which case `Thread.currentThread()` must be non-null for a real java thread. So our build matches the RI's behaviour at this point.